<img src="https://github.com/pmservice/ai-openscale-tutorials/raw/master/notebooks/images/banner.png" align="left" alt="banner">

# Working with Watson Machine Learning

This notebook should be run in a Watson Studio project, using **Default Spark 3.3.x & Python 3.10 or Python 3.9** runtime environment. **If you are viewing this in Watson Studio and do not see the required runtime env in the upper right corner of your screen, please update the runtime now.** It requires service credentials for the following Cloud services:
  * Watson OpenScale
  * Watson Machine Learning V2 plans
  * Cloud Object Storage
  
If you have a paid Cloud account, you may also provision a **Databases for PostgreSQL** or **Db2 Warehouse** service to take full advantage of integration with Watson Studio. If you choose not to provision this paid service, you can use the free internal PostgreSQL storage with OpenScale, but will not be able to configure continuous learning for your model.

The notebook will train, create and deploy a German Credit Risk model, configure OpenScale to monitor that deployment, and inject seven days' worth of historical records and measurements for viewing in the OpenScale Insights dashboard.

### Contents

- [Setup](#setup)
- [Model building and deployment](#model)
- [OpenScale configuration](#openscale)
- [Quality monitor and feedback logging](#quality)
- [Fairness, drift v2 monitoring and explanations](#fairness)
- [Custom monitors and metrics](#custom)
- [Historical data](#historical)

### Note: Samples using most latest openscale V2 client from public pypi. It does not cover followings aspects for now:

- Historical payload logging 
- Historical manual labeling

# Setup <a name="setup"></a>

## Spark check

In [ ]:
# IF you are not using  IBM Studio to run your notebook then install following packages by uncommenting it
#!pip install --upgrade pyspark==3.3.0 --no-cache | tail -n 1

In [ ]:
try:
    from pyspark.sql import SparkSession
except:
    print('Error: Spark runtime is missing. If you are using Watson Studio change the notebook runtime to Spark.')
    raise 
    
!pip list | grep ibm-watson-openscale     

ibm-watson-openscale         3.0.39

[notice] A new release of pip is available: 24.0 -> 24.1.2
[notice] To update, run: python3.10 -m pip install --upgrade pip


## Package installation

In [ ]:
# IF you are not using  IBM Studio to run your notebook then install following packages by uncommenting it

#!pip install pandas --no-cache | tail -n 1
#!pip install requests --no-cache | tail -n 1
#!pip install numpy--user --no-cache | tail -n 1
#!pip install SciPy --no-cache | tail -n 1

!pip install --upgrade ibm-watson-machine-learning --user | tail -n 1
!pip install --upgrade "ibm-watson-openscale~=3.1.2" --no-cache --user | tail -n 1

## Provision services and configure credentials

If you have not already, provision an instance of IBM Watson OpenScale using the [OpenScale link in the Cloud catalog](https://cloud.ibm.com/catalog/services/watson-openscale).

Your Cloud API key can be generated by going to the [**Users** section of the Cloud console](https://cloud.ibm.com/iam#/users). From that page, click your name, scroll down to the **API Keys** section, and click **Create an IBM Cloud API key**. Give your key a name and click **Create**, then copy the created key and paste it below.

**NOTE:** You can also get OpenScale `API_KEY` using IBM CLOUD CLI.

How to install IBM Cloud (bluemix) console: [instruction](https://console.bluemix.net/docs/cli/reference/ibmcloud/download_cli.html#install_use)

How to get api key using console:
```
bx login --sso
bx iam api-key-create 'my_key'
```

In [ ]:
CLOUD_API_KEY = ""
IAM_URL="https://iam.ng.bluemix.net/oidc/token"

If you have not already, provision an instance of IBM Watson OpenScale using the [OpenScale link in the Cloud catalog](https://cloud.ibm.com/catalog/services/watson-openscale).

Your Cloud API key can be generated by going to the [**Users** section of the Cloud console](https://cloud.ibm.com/iam#/users). From that page, click your name, scroll down to the **API Keys** section, and click **Create an IBM Cloud API key**. Give your key a name and click **Create**, then copy the created key, generate an IAM token using that key and paste it below.

### WML credentials example with API key

In [ ]:
WML_CREDENTIALS = {
                   "url": "https://us-south.ml.cloud.ibm.com",
                   "apikey": CLOUD_API_KEY
}

###  WML credentials example using IAM_token 

**NOTE**: If IAM_TOKEN is used for authentication and you receive unauthorized/expired token error at any steps, please create a new token and reinitiate clients authentication.

### Cloud object storage details

In next cells, you will need to paste some credentials to Cloud Object Storage. If you haven't worked with COS yet please visit [getting started with COS tutorial](https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-getting-started). 
You can find `COS_API_KEY_ID` and `COS_RESOURCE_CRN` variables in **_Service Credentials_** in menu of your COS instance. Used COS Service Credentials must be created with _Role_ parameter set as Writer. Later training data file will be loaded to the bucket of your instance and used as training refecence in subsription.  
`COS_ENDPOINT` variable can be found in **_Endpoint_** field of the menu.

In [ ]:
COS_API_KEY_ID = "***"
COS_RESOURCE_CRN = "***" # eg "crn:v1:bluemix:public:cloud-object-storage:global:a/3bf0d9003abfb5d29761c3e97696b71c:d6f04d83-6c4f-4a62-a165-696756d63903::"
COS_ENDPOINT = "***" # Current list avaiable at https://control.cloud-object-storage.cloud.ibm.com/v2/endpoints

In [ ]:
BUCKET_NAME = "" #example: "credit-risk-training-data"

This tutorial can use Databases for PostgreSQL, Db2 Warehouse, or a free internal verison of PostgreSQL to create a datamart for OpenScale.

If you have previously configured OpenScale, it will use your existing datamart, and not interfere with any models you are currently monitoring. Do not update the cell below.

If you do not have a paid Cloud account or would prefer not to provision this paid service, you may use the free internal PostgreSQL service with OpenScale. Do not update the cell below.

To provision a new instance of Db2 Warehouse, locate [Db2 Warehouse in the Cloud catalog](https://cloud.ibm.com/catalog/services/db2-warehouse), give your service a name, and click **Create**. Once your instance is created, click the **Service Credentials** link on the left side of the screen. Click the **New credential** button, give your credentials a name, and click **Add**. Your new credentials can be accessed by clicking the **View credentials** button. Copy and paste your Db2 Warehouse credentials into the cell below.

To provision a new instance of Databases for PostgreSQL, locate [Databases for PostgreSQL in the Cloud catalog](https://cloud.ibm.com/catalog/services/databases-for-postgresql), give your service a name, and click **Create**. Once your instance is created, click the **Service Credentials** link on the left side of the screen. Click the **New credential** button, give your credentials a name, and click **Add**. Your new credentials can be accessed by clicking the **View credentials** button. Copy and paste your Databases for PostgreSQL credentials into the cell below.

In [ ]:
DB_CREDENTIALS = None
#DB_CREDENTIALS= {"hostname":"","username":"","password":"","database":"","port":"","ssl":True,"sslmode":"","certificate_base64":""}

__If you previously configured OpenScale to use the free internal version of PostgreSQL, you can switch to a new datamart using a paid database service.__ If you would like to delete the internal PostgreSQL configuration and create a new one using service credentials supplied in the cell above, set the __KEEP_MY_INTERNAL_POSTGRES__ variable below to __False__ below. In this case, the notebook will remove your existing internal PostgreSQL datamart and create a new one with the supplied credentials. __*NO DATA MIGRATION WILL OCCUR.*__

In [ ]:
KEEP_MY_INTERNAL_POSTGRES = True

## Run the notebook

At this point, the notebook is ready to run. You can either run the cells one at a time, or click the **Kernel** option above and select **Restart and Run All** to run all the cells.

# Model building and deployment <a name="model"></a>

In this section you will learn how to train Spark MLLib model and next deploy it as web-service using Watson Machine Learning service.

## Load the training data from github

In [ ]:
from IPython.utils import io

with io.capture_output() as captured:
    !wget https://raw.githubusercontent.com/IBM/watson-openscale-samples/main/IBM%20Cloud/WML/assets/data/credit_risk/german_credit_data_biased_training.csv -O german_credit_data_biased_training.csv
!ls -lh german_credit_data_biased_training.csv

-rw-r--r--  1 nelwin  staff   673K Jul 16 11:38 german_credit_data_biased_training.csv


In [ ]:
from pyspark.sql import SparkSession
import pandas as pd
import json
import datetime

spark = SparkSession.builder.getOrCreate()
pd_data = pd.read_csv("german_credit_data_biased_training.csv", sep=",", header=0)
df_data = spark.read.csv(path="german_credit_data_biased_training.csv", sep=",", header=True, inferSchema=True)
training_data_file_name = "german_credit_data_biased_training.csv"
df_data.head()

24/07/16 11:38:14 WARN Utils: Your hostname, Nelwins-MacBook-Pro.local resolves to a loopback address: 127.0.0.1; using 192.168.29.202 instead (on interface en0)
24/07/16 11:38:14 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/07/16 11:38:15 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Row(CheckingStatus='0_to_200', LoanDuration=31, CreditHistory='credits_paid_to_date', LoanPurpose='other', LoanAmount=1889, ExistingSavings='100_to_500', EmploymentDuration='less_1', InstallmentPercent=3, Sex='female', OthersOnLoan='none', CurrentResidenceDuration=3, OwnsProperty='savings_insurance', Age=32, InstallmentPlans='none', Housing='own', ExistingCreditsCount=1, Job='skilled', Dependents=1, Telephone='none', ForeignWorker='yes', Risk='No Risk')

## Explore data

In [ ]:
df_data.printSchema()

root
 |-- CheckingStatus: string (nullable = true)
 |-- LoanDuration: integer (nullable = true)
 |-- CreditHistory: string (nullable = true)
 |-- LoanPurpose: string (nullable = true)
 |-- LoanAmount: integer (nullable = true)
 |-- ExistingSavings: string (nullable = true)
 |-- EmploymentDuration: string (nullable = true)
 |-- InstallmentPercent: integer (nullable = true)
 |-- Sex: string (nullable = true)
 |-- OthersOnLoan: string (nullable = true)
 |-- CurrentResidenceDuration: integer (nullable = true)
 |-- OwnsProperty: string (nullable = true)
 |-- Age: integer (nullable = true)
 |-- InstallmentPlans: string (nullable = true)
 |-- Housing: string (nullable = true)
 |-- ExistingCreditsCount: integer (nullable = true)
 |-- Job: string (nullable = true)
 |-- Dependents: integer (nullable = true)
 |-- Telephone: string (nullable = true)
 |-- ForeignWorker: string (nullable = true)
 |-- Risk: string (nullable = true)



In [ ]:
print("Number of records: " + str(df_data.count()))

Number of records: 5000


## Save training data to Cloud Object Storage

In [ ]:
import ibm_boto3
from ibm_botocore.client import Config, ClientError

cos_client = ibm_boto3.resource("s3",
    ibm_api_key_id=COS_API_KEY_ID,
    ibm_service_instance_id=COS_RESOURCE_CRN,
    ibm_auth_endpoint="https://iam.bluemix.net/oidc/token",
    config=Config(signature_version="oauth"),
    endpoint_url=COS_ENDPOINT
)

In [ ]:
with open(training_data_file_name, "rb") as file_data:
    cos_client.Object(BUCKET_NAME, training_data_file_name).upload_fileobj(
        Fileobj=file_data
    )

## Create a model

In [ ]:
spark_df = df_data
(train_data, test_data) = spark_df.randomSplit([0.8, 0.2], 24)

MODEL_NAME = "Spark German Risk Model - Final"
DEPLOYMENT_NAME = "Spark German Risk Deployment - Final"

print("Number of records for training: " + str(train_data.count()))
print("Number of records for evaluation: " + str(test_data.count()))

Number of records for training: 4005
Number of records for evaluation: 995


The code below creates a Random Forest Classifier with Spark, setting up string indexers for the categorical features and the label column. Finally, this notebook creates a pipeline including the indexers and the model, and does an initial Area Under ROC evaluation of the model.

In [ ]:
from pyspark.ml.feature import OneHotEncoder, StringIndexer, IndexToString, VectorAssembler
from pyspark.ml.evaluation import BinaryClassificationEvaluator
from pyspark.ml import Pipeline, Model
from pyspark.ml.feature import SQLTransformer

features = [x for x in spark_df.columns if x != 'Risk']
categorical_features = ['CheckingStatus', 'CreditHistory', 'LoanPurpose', 'ExistingSavings', 'EmploymentDuration', 'Sex', 'OthersOnLoan', 'OwnsProperty', 'InstallmentPlans', 'Housing', 'Job', 'Telephone', 'ForeignWorker']
categorical_num_features = [x + '_IX' for x in categorical_features]
si_list = [StringIndexer(inputCol=x, outputCol=y) for x, y in zip(categorical_features, categorical_num_features)]
va_features = VectorAssembler(inputCols=categorical_num_features + [x for x in features if x not in categorical_features], outputCol="features")

In [ ]:
si_label = StringIndexer(inputCol="Risk", outputCol="label").fit(spark_df)
label_converter = IndexToString(inputCol="prediction", outputCol="predictedLabel", labels=si_label.labels)

In [ ]:
from pyspark.ml.classification import RandomForestClassifier

classifier = RandomForestClassifier(featuresCol="features")
feature_filter = SQLTransformer(statement="SELECT * FROM __THIS__")
pipeline = Pipeline(stages= si_list + [si_label, va_features, classifier, label_converter, feature_filter])
model_rf = pipeline.fit(train_data)

**Note:** If you want filter features from model output please replace **`*`** with feature names to be retained in **`SQLTransformer`** statement.

In [ ]:
predictions = model_rf.transform(test_data)
evaluatorDT = BinaryClassificationEvaluator(rawPredictionCol="prediction")
area_under_curve = evaluatorDT.evaluate(predictions)

print("areaUnderROC = %g" % area_under_curve)

24/07/16 11:38:31 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


areaUnderROC = 0.722391


## Publish the model

In this section, the notebook uses the supplied Watson Machine Learning credentials to save the model (including the pipeline) to the WML instance. Previous versions of the model are removed so that the notebook can be run again, resetting all data for another demo.

In [ ]:
import json
from ibm_watson_machine_learning import APIClient

wml_client = APIClient(WML_CREDENTIALS)
wml_client.version

'1.0.360'

### Listing all the available spaces

In [ ]:
wml_client.spaces.list(limit=10)

------------------------------------  -------------------------------------------------------------------  ------------------------
ID                                    NAME                                                                 CREATED
03b71762-7fcd-b2d8-b36d-4dab3e246e14  Automotive Demo - Quality report summarization                       2024-05-27T13:13:27.233Z
0e288639-8e0e-f191-bcbd-cae000d2d40b  pre-prod-space                                                       2024-02-28T07:35:30.368Z
ad31d6ea-c8a6-68ce-55f0-0a1ba7a738a9  prompt-space                                                         2024-01-15T19:29:21.535Z
2ce5fbc6-d167-9ba7-f9fe-1bc940ece879  llm. space                                                           2023-11-24T13:57:05.167Z
bc41b128-2d09-4e8a-07cc-def05be623e1  wml                                                                  2023-09-21T07:06:42.577Z
8ce89215-5ddd-7dd0-1318-dd083a78c24c  AutoAIDemo                                             

,ID,NAME,CREATED
0,03b71762-7fcd-b2d8-b36d-4dab3e246e14,Automotive Demo - Quality report summarization,2024-05-27T13:13:27.233Z
1,0e288639-8e0e-f191-bcbd-cae000d2d40b,pre-prod-space,2024-02-28T07:35:30.368Z
2,ad31d6ea-c8a6-68ce-55f0-0a1ba7a738a9,prompt-space,2024-01-15T19:29:21.535Z
3,2ce5fbc6-d167-9ba7-f9fe-1bc940ece879,llm. space,2023-11-24T13:57:05.167Z
4,bc41b128-2d09-4e8a-07cc-def05be623e1,wml,2023-09-21T07:06:42.577Z
5,8ce89215-5ddd-7dd0-1318-dd083a78c24c,AutoAIDemo,2023-08-25T02:50:27.113Z
6,2743a4fe-f016-32fc-a8fd-d96e6fd1ad72,openscale-express-path-preprod-80e6093f-5acf-4...,2023-08-16T06:59:18.538Z
7,05446376-430b-4c43-f5ca-6b87e5ef17e9,openscale-express-path-80e6093f-5acf-4eb7-9da6...,2023-08-16T06:58:57.332Z


In [ ]:
WML_SPACE_ID='' # use space id here
wml_client.set.default_space(WML_SPACE_ID)

'SUCCESS'

### Remove existing model and deployment

In [ ]:
deployments_list = wml_client.deployments.get_details()
for deployment in deployments_list["resources"]:
    model_id = deployment["entity"]["asset"]["id"]
    deployment_id = deployment["metadata"]["id"]
    if deployment["metadata"]["name"] == DEPLOYMENT_NAME:
        print("Deleting deployment id", deployment_id)
        wml_client.deployments.delete(deployment_id)
        print("Deleting model id", model_id)
        wml_client.repository.delete(model_id)
wml_client.repository.list_models()

------------------------------------  ----------------------------------------  ------------------------  ---------  ----------  ----------------
ID                                    NAME                                      CREATED                   TYPE       SPEC_STATE  SPEC_REPLACEMENT
fdae2ba2-7346-92ab-12ef-984e779d0938  Spark German Risk Model - Training stats  2024-07-15T10:04:40.002Z  mllib_3.3  supported
------------------------------------  ----------------------------------------  ------------------------  ---------  ----------  ----------------


,ID,NAME,CREATED,TYPE,SPEC_STATE,SPEC_REPLACEMENT
0,fdae2ba2-7346-92ab-12ef-984e779d0938,Spark German Risk Model - Training stats,2024-07-15T10:04:40.002Z,mllib_3.3,supported,


In [ ]:
#Create connections to a COS bucket
datasource_type = wml_client.connections.get_datasource_type_uid_by_name('bluemixcloudobjectstorage')
conn_meta_props= {
    wml_client.connections.ConfigurationMetaNames.NAME: "Connection My COS ",
    wml_client.connections.ConfigurationMetaNames.DATASOURCE_TYPE: datasource_type,
    wml_client.connections.ConfigurationMetaNames.DESCRIPTION: "Connection to my COS",
    wml_client.connections.ConfigurationMetaNames.PROPERTIES: {
        'bucket': BUCKET_NAME,
        'api_key': CLOUD_API_KEY,
        'resource_instance_id': COS_RESOURCE_CRN,
        'iam_url': IAM_URL,
        'url': COS_ENDPOINT
    }
}

conn_details = wml_client.connections.create(meta_props=conn_meta_props)
connection_id = wml_client.connections.get_uid(conn_details)
connection_id

Creating connections...
SUCCESS


'cc296828-3b80-712c-2938-c999b2e42ba1'

In [ ]:
training_data_references = [
                {
                    "id": "German Credit Risk", 
                    "type": "connection_asset",
                    "connection": {
                        "id": connection_id,
                        "href": "/v2/connections/" + connection_id + "?space_id=" + WML_SPACE_ID
                        
                    },
                    "location": {
                        "bucket": BUCKET_NAME,
                        "file_name": training_data_file_name
                    },
                    "schema": {
                    "id": "training_schema",
                    "fields": [
                      {
                        "metadata": {},
                        "name": "CheckingStatus",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "LoanDuration",
                        "nullable": True,
                        "type": "integer"
                      },
                      {
                        "metadata": {},
                        "name": "CreditHistory",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "LoanPurpose",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "LoanAmount",
                        "nullable": True,
                        "type": "integer"
                      },
                      {
                        "metadata": {},
                        "name": "ExistingSavings",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "EmploymentDuration",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "InstallmentPercent",
                        "nullable": True,
                        "type": "integer"
                      },
                      {
                        "metadata": {},
                        "name": "Sex",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "OthersOnLoan",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "CurrentResidenceDuration",
                        "nullable": True,
                        "type": "integer"
                      },
                      {
                        "metadata": {},
                        "name": "OwnsProperty",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "Age",
                        "nullable": True,
                        "type": "integer"
                      },
                      {
                        "metadata": {},
                        "name": "InstallmentPlans",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "Housing",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "ExistingCreditsCount",
                        "nullable": True,
                        "type": "integer"
                      },
                      {
                        "metadata": {},
                        "name": "Job",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "Dependents",
                        "nullable": True,
                        "type": "integer"
                      },
                      {
                        "metadata": {},
                        "name": "Telephone",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {},
                        "name": "ForeignWorker",
                        "nullable": True,
                        "type": "string"
                      },
                      {
                        "metadata": {
                          "modeling_role": "target"
                        },
                        "name": "Risk",
                        "nullable": True,
                        "type": "string"
                      }
                    ]
                  }
                }
            ]

In [ ]:
software_spec_uid = wml_client.software_specifications.get_id_by_name("spark-mllib_3.3")
print("Software Specification ID: {}".format(software_spec_uid))
model_props = {
        wml_client._models.ConfigurationMetaNames.NAME:"{}".format(MODEL_NAME),
        wml_client._models.ConfigurationMetaNames.TYPE: "mllib_3.3",
        wml_client._models.ConfigurationMetaNames.SOFTWARE_SPEC_UID: software_spec_uid,
        wml_client._models.ConfigurationMetaNames.TRAINING_DATA_REFERENCES: training_data_references,
        wml_client._models.ConfigurationMetaNames.LABEL_FIELD: "Risk",
    }

Software Specification ID: 6b32b603-97d7-891d-ef8d-e687d7d86d19


In [ ]:
print("Storing model ...")
published_model_details = wml_client.repository.store_model(
    model=model_rf, 
    meta_props=model_props, 
    training_data=train_data, 
    pipeline=pipeline)

model_uid = wml_client.repository.get_model_id(published_model_details)
print("Done")
print("Model ID: {}".format(model_uid))

Storing model ...
Done
Model ID: 857af6d0-2c26-572a-7430-277e506d0372


## Deploy the model

The next section of the notebook deploys the model as a RESTful web service in Watson Machine Learning. The deployed model will have a scoring URL you can use to send data to the model for predictions.

In [ ]:
deployment_details = wml_client.deployments.create(
    model_uid, 
    meta_props={
        wml_client.deployments.ConfigurationMetaNames.NAME: "{}".format(DEPLOYMENT_NAME),
        wml_client.deployments.ConfigurationMetaNames.ONLINE: {}
    }
)
scoring_url = wml_client.deployments.get_scoring_href(deployment_details)
deployment_uid=wml_client.deployments.get_uid(deployment_details)

print("Scoring URL:" + scoring_url)
print("Model id: {}".format(model_uid))
print("Deployment id: {}".format(deployment_uid))



#######################################################################################

Synchronous deployment creation for uid: '857af6d0-2c26-572a-7430-277e506d0372' started

#######################################################################################


initializing
Note: online_url and serving_urls are deprecated and will be removed in a future release. Use inference instead.
.
ready


------------------------------------------------------------------------------------------------
Successfully finished deployment creation, deployment_uid='198d218d-a4d7-88c9-433a-2cbe9b9ef9ef'
------------------------------------------------------------------------------------------------


Scoring URL:https://us-south.ml.cloud.ibm.com/ml/v4/deployments/198d218d-a4d7-88c9-433a-2cbe9b9ef9ef/predictions
Model id: 857af6d0-2c26-572a-7430-277e506d0372
Deployment id: 198d218d-a4d7-88c9-433a-2cbe9b9ef9ef


## Sample scoring

In [ ]:
fields = ["CheckingStatus", "LoanDuration", "CreditHistory", "LoanPurpose", "LoanAmount", "ExistingSavings",
                  "EmploymentDuration", "InstallmentPercent", "Sex", "OthersOnLoan", "CurrentResidenceDuration",
                  "OwnsProperty", "Age", "InstallmentPlans", "Housing", "ExistingCreditsCount", "Job", "Dependents",
                  "Telephone", "ForeignWorker"]
values = [
            ["no_checking", 13, "credits_paid_to_date", "car_new", 1343, "100_to_500", "1_to_4", 2, "female", "none", 3,
             "savings_insurance", 46, "none", "own", 2, "skilled", 1, "none", "yes"],
            ["no_checking", 24, "prior_payments_delayed", "furniture", 4567, "500_to_1000", "1_to_4", 4, "male", "none",
             4, "savings_insurance", 36, "none", "free", 2, "management_self-employed", 1, "none", "yes"],
        ]

scoring_payload = {"input_data": [{"fields": fields, "values": values}]}

In [ ]:
scoring_response = wml_client.deployments.score(deployment_uid, scoring_payload)
scoring_response

{'predictions': [{'fields': ['CheckingStatus',
    'LoanDuration',
    'CreditHistory',
    'LoanPurpose',
    'LoanAmount',
    'ExistingSavings',
    'EmploymentDuration',
    'InstallmentPercent',
    'Sex',
    'OthersOnLoan',
    'CurrentResidenceDuration',
    'OwnsProperty',
    'Age',
    'InstallmentPlans',
    'Housing',
    'ExistingCreditsCount',
    'Job',
    'Dependents',
    'Telephone',
    'ForeignWorker',
    'CheckingStatus_IX',
    'CreditHistory_IX',
    'LoanPurpose_IX',
    'ExistingSavings_IX',
    'EmploymentDuration_IX',
    'Sex_IX',
    'OthersOnLoan_IX',
    'OwnsProperty_IX',
    'InstallmentPlans_IX',
    'Housing_IX',
    'Job_IX',
    'Telephone_IX',
    'ForeignWorker_IX',
    'features',
    'rawPrediction',
    'probability',
    'prediction',
    'predictedLabel'],
   'values': [['no_checking',
     13,
     'credits_paid_to_date',
     'car_new',
     1343,
     '100_to_500',
     '1_to_4',
     2,
     'female',
     'none',
     3,
     'savings

# Configure OpenScale <a name="openscale"></a>

The notebook will now import the necessary libraries and set up a Python OpenScale client.

In [ ]:
from ibm_cloud_sdk_core.authenticators import IAMAuthenticator,BearerTokenAuthenticator

from ibm_watson_openscale import *
from ibm_watson_openscale.supporting_classes.enums import *
from ibm_watson_openscale.supporting_classes import *


authenticator = IAMAuthenticator(apikey=CLOUD_API_KEY)
#authenticator = BearerTokenAuthenticator(bearer_token=IAM_TOKEN) ## uncomment this line if using IAM token to authenticate
wos_client = APIClient(authenticator=authenticator)
wos_client.version

'3.0.39'

## Create schema and datamart

### Set up datamart

Watson OpenScale uses a database to store payload logs and calculated metrics. If database credentials were **not** supplied above, the notebook will use the free, internal lite database. If database credentials were supplied, the datamart will be created there **unless** there is an existing datamart **and** the **KEEP_MY_INTERNAL_POSTGRES** variable is set to **True**. If an OpenScale datamart exists in Db2 or PostgreSQL, the existing datamart will be used and no data will be overwritten.

Prior instances of the German Credit model will be removed from OpenScale monitoring.

In [ ]:
wos_client.data_marts.show()

AIOSFASTPATH-8d99dfc4-eaea-8620-10c7-36a38f2f31c8,None,True,active,2024-05-16 06:09:07.089000+00:00,80e6093f-5acf-4eb7-9da6-7ba9bf56a929


In [ ]:
data_marts = wos_client.data_marts.list().result.data_marts
if len(data_marts) == 0:
    if DB_CREDENTIALS is not None:
        if SCHEMA_NAME is None: 
            print("Please specify the SCHEMA_NAME and rerun the cell")

        print('Setting up external datamart')
        added_data_mart_result = wos_client.data_marts.add(
                background_mode=False,
                name="WOS Data Mart",
                description="Data Mart created by WOS tutorial notebook",
                database_configuration=DatabaseConfigurationRequest(
                  database_type=DatabaseType.POSTGRESQL,
                    credentials=PrimaryStorageCredentialsLong(
                        hostname=DB_CREDENTIALS['hostname'],
                        username=DB_CREDENTIALS['username'],
                        password=DB_CREDENTIALS['password'],
                        db=DB_CREDENTIALS['database'],
                        port=DB_CREDENTIALS['port'],
                        ssl=True,
                        sslmode=DB_CREDENTIALS['sslmode'],
                        certificate_base64=DB_CREDENTIALS['certificate_base64']
                    ),
                    location=LocationSchemaName(
                        schema_name= SCHEMA_NAME
                    )
                )
             ).result
    else:
        print('Setting up internal datamart')
        added_data_mart_result = wos_client.data_marts.add(
                background_mode=False,
                name="WOS Data Mart",
                description="Data Mart created by WOS tutorial notebook", 
                internal_database = True).result
        
    data_mart_id = added_data_mart_result.metadata.id
    
else:
    data_mart_id=data_marts[0].metadata.id
    print('Using existing datamart {}'.format(data_mart_id))

Using existing datamart 80e6093f-5acf-4eb7-9da6-7ba9bf56a929


### Remove existing service provider connected with used  WML instance. 

Multiple service providers for the same engine instance are avaiable in Watson OpenScale. To avoid multiple service providers of used WML instance in the tutorial notebook the following code deletes existing service provder(s) and then adds new one. 

In [ ]:
SERVICE_PROVIDER_NAME = "Watson Machine Learning V2"
SERVICE_PROVIDER_DESCRIPTION = "Added by tutorial WOS notebook."

In [ ]:
service_providers = wos_client.service_providers.list().result.service_providers
for service_provider in service_providers:
    service_instance_name = service_provider.entity.name
    if service_instance_name == SERVICE_PROVIDER_NAME:
        service_provider_id = service_provider.metadata.id
        wos_client.service_providers.delete(service_provider_id)
        print("Deleted existing service_provider for WML instance: {}".format(service_provider_id))

## Add service provider

Watson OpenScale needs to be bound to the Watson Machine Learning instance to capture payload data into and out of the model.

**Note:** You can bind more than one engine instance if needed by calling `wos_client.service_providers.add` method. Next, you can refer to particular service provider using `service_provider_id`.

In [ ]:
added_service_provider_result = wos_client.service_providers.add(
        name=SERVICE_PROVIDER_NAME,
        description=SERVICE_PROVIDER_DESCRIPTION,
        service_type=ServiceTypes.WATSON_MACHINE_LEARNING,
        deployment_space_id = WML_SPACE_ID,
        operational_space_id = "production",
        credentials=WMLCredentialsCloud(
            apikey=CLOUD_API_KEY,      ## use `apikey=IAM_TOKEN` if using IAM_TOKEN to initiate client
            url=WML_CREDENTIALS["url"],
            instance_id=None
        ),
        background_mode=False
    ).result
service_provider_id = added_service_provider_result.metadata.id




 Waiting for end of adding service provider ba3fa5df-a631-ff60-9d4c-9a36ce63408d 




active

-----------------------------------------------
 Successfully finished adding service provider 
-----------------------------------------------




In [ ]:
wos_client.service_providers.show()

19e7e320-bb3e-b012-2f72-62c1c54ffade,active,Watson Machine Learning V2-2,watson_machine_learning,2024-07-16 06:11:13.858000+00:00,ba3fa5df-a631-ff60-9d4c-9a36ce63408d
None,active,Azure Machine Learning,azure_machine_learning,2024-07-15 11:18:22.795000+00:00,94eff0d0-dcc8-4994-b64a-0bbce1dea7e8
None,active,AWS Machine Learning,amazon_sagemaker,2024-07-11 09:29:39.272000+00:00,035dd138-ec3b-7fdc-3586-153a846b6676
None,active,OpenScale Headless Service Provider,custom_machine_learning,2024-07-05 13:55:20.532000+00:00,3ae64973-e61e-8e3a-9cd7-bdea8d94557e
19e7e320-bb3e-b012-2f72-62c1c54ffade,active,aWatson Machine Learning V2_amzon,watson_machine_learning,2024-06-06 06:03:03.459000+00:00,182685ee-c8f9-cd05-4fbe-0e051e1600fe
19e7e320-bb3e-b012-2f72-62c1c54ffade,active,Watson Machine Learning V2,watson_machine_learning,2024-06-02 08:24:57.362000+00:00,53fc0337-4be3-1e49-d447-5289d3e63450
19e7e320-bb3e-b012-2f72-62c1c54ffade,active,WOS ExpressPath WML pre_production binding,watson_machine_learning,2024-05-16 06:09:28.464000+00:00,f1ffce7c-9148-e8d4-a07b-3fac97bb490e
19e7e320-bb3e-b012-2f72-62c1c54ffade,active,WOS ExpressPath WML production binding,watson_machine_learning,2024-05-16 06:09:12.960000+00:00,3767c195-2419-f7fe-51bc-72bcd018abd4


In [ ]:
asset_deployment_details = wos_client.service_providers.list_assets(data_mart_id=data_mart_id, service_provider_id=service_provider_id, deployment_space_id = WML_SPACE_ID).result['resources'][0]
asset_deployment_details

{'metadata': {'guid': '0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644',
  'url': 'https://us-south.ml.cloud.ibm.com/ml/v4/deployments/0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644?space_id=ad31d6ea-c8a6-68ce-55f0-0a1ba7a738a9',
  'created_at': '2024-07-15T10:04:51.382Z',
  'modified_at': '2024-07-15T10:04:51.382Z'},
 'entity': {'name': 'Spark German Risk Deployment - Training stats',
  'type': 'online',
  'scoring_endpoint': {'url': 'https://us-south.ml.cloud.ibm.com/ml/v4/deployments/0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644/predictions'},
  'asset': {},
  'asset_properties': {}}}

In [ ]:
model_asset_details_from_deployment=wos_client.service_providers.get_deployment_asset(data_mart_id=data_mart_id,service_provider_id=service_provider_id,deployment_id=deployment_uid,deployment_space_id=WML_SPACE_ID)
model_asset_details_from_deployment

## Subscriptions

### Remove existing credit risk subscriptions

This code removes previous subscriptions to the German Credit model to refresh the monitors with the new model and new data.

In [ ]:
wos_client.subscriptions.show()

085460ef94636166aea5800e9ea26168,model,GermanCreditRisk.2019.1.9.10.41.58.611,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,563f01d37f720857b95c557dc76176ad,GermanCreditRisk.2019.1.9.10.41.58.611,94eff0d0-dcc8-4994-b64a-0bbce1dea7e8,active,2024-07-15 11:20:20.209000+00:00,07d82ce8-8936-4a13-3261-392d9239aa04
98783e76b0aedefc06c70bd74fa2356e,model,Credit-risk-linear-learner-2021-06-24-05-32,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,cc7acf711b19c689d67e7b2a01b88ca2,Credit-risk-endpoint-scoring-2021-06-24-05-32,035dd138-ec3b-7fdc-3586-153a846b6676,active,2024-07-11 09:30:10.069000+00:00,8dd6a05c-5ea3-8e9b-0d0c-08e200fc9dbc
0abd8a38-6827-1821-88d7-3bdc3906c683,model,[asset] GCR Headless Subscription,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,8b75918d-3f03-34cb-5b68-798b21e563ff,[asset] GCR Headless Subscription,3ae64973-e61e-8e3a-9cd7-bdea8d94557e,active,2024-07-05 13:57:01.523000+00:00,e223ca0c-e6f4-5074-21a3-1a3b986ff0eb
25458fae-e6f8-d0b9-a78c-fd2b99ccae59,model,Spark German Risk Model - Final,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,5d9d0108-adb4-0254-beae-bd628a023faf,New - Spark German Risk Deployment - Final,53fc0337-4be3-1e49-d447-5289d3e63450,active,2024-07-04 09:15:28.269000+00:00,344f0435-659d-9b60-1f91-5dda0e785987
04f40aeb-63f7-e7d6-6108-0b3814eb155f,model,Amazon Mobile Review Classifier,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,b8bb3373-2487-36db-8d37-11087ca456d6,Amazon Mobile Review Classifier deployment2,182685ee-c8f9-cd05-4fbe-0e051e1600fe,active,2024-06-06 06:04:06.732000+00:00,695a04bf-402a-31c2-6a5d-6a2ff4a350c4
ce277025-5bbf-6037-eb15-60ad3ba8d1c9,model,Scikit German Risk Model WML V4,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,86b4991a-9d93-3718-66a7-874cfd8114df,Scikit German Risk Deployment WML V4-1,53fc0337-4be3-1e49-d447-5289d3e63450,active,2024-06-02 08:26:51.212000+00:00,686e4ec1-2b12-cd71-ba08-9e1087124a7b
af939ffe-15b7-3b17-5ede-e005d610f357,model,GermanCreditRiskModelChallenger,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,44b2b767-e81a-8870-01cb-09da416d55b8,GermanCreditRiskModelChallenger,f1ffce7c-9148-e8d4-a07b-3fac97bb490e,active,2024-05-16 06:09:55.832000+00:00,35028a56-b784-7a4e-01eb-e268c83a977d


This code removes previous subscriptions to the German Credit model to refresh the monitors with the new model and new data.

In [ ]:
subscriptions = wos_client.subscriptions.list().result.subscriptions
for subscription in subscriptions:
    sub_model_id = subscription.entity.asset.asset_id
    if sub_model_id == model_uid:
        wos_client.subscriptions.delete(subscription.metadata.id)
        print('Deleted existing subscription for model', sub_model_id)

This code creates the model subscription in OpenScale using the Python client API. Note that we need to provide the model unique identifier, and some information about the model itself.

### This code creates the model subscription in OpenScale using the Python client API. Note that we need to provide the model unique identifier, and some information about the model itself.

Note :
1) User can add `training_data_schema` as an optional parameter as part of  AssetPropertiesRequest while creating subscription.
2) If not added, schemas will get added/updated after first payload logging

In [ ]:
# training_data_schema=SparkStruct.from_dict(model_asset_details_from_deployment["entity"]["asset_properties"]["training_data_schema"]

In [ ]:
from ibm_watson_openscale.base_classes.watson_open_scale_v2 import ScoringEndpointRequest

In [ ]:
subscription_details = wos_client.subscriptions.add(
        background_mode=False,
        data_mart_id=data_mart_id,
        service_provider_id=service_provider_id,
        asset=Asset(
            asset_id=model_asset_details_from_deployment["entity"]["asset"]["asset_id"],
            name=model_asset_details_from_deployment["entity"]["asset"]["name"],
            url=model_asset_details_from_deployment["entity"]["asset"]["url"],
            asset_type=AssetTypes.MODEL,
            input_data_type=InputDataType.STRUCTURED,
            problem_type=ProblemType.BINARY_CLASSIFICATION
        ),
        deployment=AssetDeploymentRequest(
            deployment_id=asset_deployment_details['metadata']['guid'],
            name=asset_deployment_details['entity']['name'],
            deployment_type= DeploymentTypes.ONLINE,
            url=asset_deployment_details['metadata']['url'],
            scoring_endpoint=ScoringEndpointRequest(url=scoring_url) # score model without shadow deployment
        ),
        asset_properties=AssetPropertiesRequest(
            label_column='Risk',
            probability_fields=['probability'],
            prediction_field='predictedLabel',
            feature_fields = ["CheckingStatus","LoanDuration","CreditHistory","LoanPurpose","LoanAmount","ExistingSavings","EmploymentDuration","InstallmentPercent","Sex","OthersOnLoan","CurrentResidenceDuration","OwnsProperty","Age","InstallmentPlans","Housing","ExistingCreditsCount","Job","Dependents","Telephone","ForeignWorker"],
            categorical_fields = ["CheckingStatus","CreditHistory","LoanPurpose","ExistingSavings","EmploymentDuration","Sex","OthersOnLoan","OwnsProperty","InstallmentPlans","Housing","Job","Telephone","ForeignWorker"],
            training_data_reference=TrainingDataReference(type='cos',
                                                          location=COSTrainingDataReferenceLocation(bucket = BUCKET_NAME,
                                                                                                    file_name = training_data_file_name),
                                                          connection=COSTrainingDataReferenceConnection.from_dict({
                                                                        "resource_instance_id": COS_RESOURCE_CRN,
                                                                        "url": COS_ENDPOINT,
                                                                        "api_key": COS_API_KEY_ID,
                                                                        "iam_url": IAM_URL}))
        )
    ).result
subscription_id = subscription_details.metadata.id
subscription_id




 Waiting for end of adding subscription 3a293f0a-318d-ce62-423a-9376b02b966a 




active

-------------------------------------------
 Successfully finished adding subscription 
-------------------------------------------




'3a293f0a-318d-ce62-423a-9376b02b966a'

In [ ]:
import time

time.sleep(5)
payload_data_set_id = None
payload_data_set_id = wos_client.data_sets.list(type=DataSetTypes.PAYLOAD_LOGGING, 
                                                target_target_id=subscription_id, 
                                                target_target_type=TargetTypes.SUBSCRIPTION).result.data_sets[0].metadata.id
if payload_data_set_id is None:
    print("Payload data set not found. Please check subscription status.")
else:
    print("Payload data set id: ", payload_data_set_id)

Payload data set id:  81c3bd30-c151-a544-b737-476606a2c992


In [ ]:
wos_client.data_sets.show()

80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,3a293f0a-318d-ce62-423a-9376b02b966a,subscription,model_health,2024-07-16 06:11:34.570000+00:00,93727936-0f25-66a4-ba4d-3742c52336d9
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,3a293f0a-318d-ce62-423a-9376b02b966a,subscription,payload_logging_error,2024-07-16 06:11:34.305000+00:00,6ce4771e-56dc-3bed-b5ed-bc418b0ad5c7
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,3a293f0a-318d-ce62-423a-9376b02b966a,subscription,manual_labeling,2024-07-16 06:11:34.024000+00:00,bd47b5de-a3fe-4b6d-bb7f-7f61e074cf3c
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,3a293f0a-318d-ce62-423a-9376b02b966a,subscription,payload_logging,2024-07-16 06:11:33.622000+00:00,81c3bd30-c151-a544-b737-476606a2c992
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,07d82ce8-8936-4a13-3261-392d9239aa04,subscription,explanations,2024-07-15 11:22:39.237000+00:00,04f40d43-6209-37f7-41e7-f0a429f0fd26
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,07d82ce8-8936-4a13-3261-392d9239aa04,subscription,model_health,2024-07-15 11:20:23.058000+00:00,342be89d-3682-bcbb-8a2b-b2101e932843
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,07d82ce8-8936-4a13-3261-392d9239aa04,subscription,feedback,2024-07-15 11:21:08.826000+00:00,4b1b0c87-a9e4-249c-f9f1-6a0aa5e10e7f
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,07d82ce8-8936-4a13-3261-392d9239aa04,subscription,payload_logging,2024-07-15 11:20:21.418000+00:00,7b310cc9-0548-80bc-326a-d1d9ae297ae4
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,07d82ce8-8936-4a13-3261-392d9239aa04,subscription,payload_logging_error,2024-07-15 11:20:21.915000+00:00,3da0d056-e1ab-df84-844f-18c2ecdcb7f6
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,07d82ce8-8936-4a13-3261-392d9239aa04,subscription,manual_labeling,2024-07-15 11:20:21.667000+00:00,9a8d5e07-5dee-d487-7cc6-a8b172236233


Note: First 10 records were displayed.


Get subscription list

In [ ]:
wos_client.subscriptions.show()

857af6d0-2c26-572a-7430-277e506d0372,model,Spark German Risk Model - Final,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644,Spark German Risk Deployment - Training stats,ba3fa5df-a631-ff60-9d4c-9a36ce63408d,active,2024-07-16 06:11:32.368000+00:00,3a293f0a-318d-ce62-423a-9376b02b966a
085460ef94636166aea5800e9ea26168,model,GermanCreditRisk.2019.1.9.10.41.58.611,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,563f01d37f720857b95c557dc76176ad,GermanCreditRisk.2019.1.9.10.41.58.611,94eff0d0-dcc8-4994-b64a-0bbce1dea7e8,active,2024-07-15 11:20:20.209000+00:00,07d82ce8-8936-4a13-3261-392d9239aa04
98783e76b0aedefc06c70bd74fa2356e,model,Credit-risk-linear-learner-2021-06-24-05-32,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,cc7acf711b19c689d67e7b2a01b88ca2,Credit-risk-endpoint-scoring-2021-06-24-05-32,035dd138-ec3b-7fdc-3586-153a846b6676,active,2024-07-11 09:30:10.069000+00:00,8dd6a05c-5ea3-8e9b-0d0c-08e200fc9dbc
0abd8a38-6827-1821-88d7-3bdc3906c683,model,[asset] GCR Headless Subscription,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,8b75918d-3f03-34cb-5b68-798b21e563ff,[asset] GCR Headless Subscription,3ae64973-e61e-8e3a-9cd7-bdea8d94557e,active,2024-07-05 13:57:01.523000+00:00,e223ca0c-e6f4-5074-21a3-1a3b986ff0eb
25458fae-e6f8-d0b9-a78c-fd2b99ccae59,model,Spark German Risk Model - Final,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,5d9d0108-adb4-0254-beae-bd628a023faf,New - Spark German Risk Deployment - Final,53fc0337-4be3-1e49-d447-5289d3e63450,active,2024-07-04 09:15:28.269000+00:00,344f0435-659d-9b60-1f91-5dda0e785987
04f40aeb-63f7-e7d6-6108-0b3814eb155f,model,Amazon Mobile Review Classifier,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,b8bb3373-2487-36db-8d37-11087ca456d6,Amazon Mobile Review Classifier deployment2,182685ee-c8f9-cd05-4fbe-0e051e1600fe,active,2024-06-06 06:04:06.732000+00:00,695a04bf-402a-31c2-6a5d-6a2ff4a350c4
ce277025-5bbf-6037-eb15-60ad3ba8d1c9,model,Scikit German Risk Model WML V4,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,86b4991a-9d93-3718-66a7-874cfd8114df,Scikit German Risk Deployment WML V4-1,53fc0337-4be3-1e49-d447-5289d3e63450,active,2024-06-02 08:26:51.212000+00:00,686e4ec1-2b12-cd71-ba08-9e1087124a7b
af939ffe-15b7-3b17-5ede-e005d610f357,model,GermanCreditRiskModelChallenger,80e6093f-5acf-4eb7-9da6-7ba9bf56a929,44b2b767-e81a-8870-01cb-09da416d55b8,GermanCreditRiskModelChallenger,f1ffce7c-9148-e8d4-a07b-3fac97bb490e,active,2024-05-16 06:09:55.832000+00:00,35028a56-b784-7a4e-01eb-e268c83a977d


### Score the model so we can configure monitors

Now that the WML service has been bound and the subscription has been created, we need to send a request to the model before we configure OpenScale. This allows OpenScale to create a payload log in the datamart with the correct schema, so it can capture data coming into and out of the model. First, the code gets the model deployment's endpoint URL, and then sends a few records for predictions.

In [ ]:
fields = ["CheckingStatus","LoanDuration","CreditHistory","LoanPurpose","LoanAmount","ExistingSavings","EmploymentDuration","InstallmentPercent","Sex","OthersOnLoan","CurrentResidenceDuration","OwnsProperty","Age","InstallmentPlans","Housing","ExistingCreditsCount","Job","Dependents","Telephone","ForeignWorker"]
values = [
  ["no_checking",13,"credits_paid_to_date","car_new",1343,"100_to_500","1_to_4",2,"female","none",3,"savings_insurance",46,"none","own",2,"skilled",1,"none","yes"],
  ["no_checking",24,"prior_payments_delayed","furniture",4567,"500_to_1000","1_to_4",4,"male","none",4,"savings_insurance",36,"none","free",2,"management_self-employed",1,"none","yes"],
  ["0_to_200",26,"all_credits_paid_back","car_new",863,"less_100","less_1",2,"female","co-applicant",2,"real_estate",38,"none","own",1,"skilled",1,"none","yes"],
  ["0_to_200",14,"no_credits","car_new",2368,"less_100","1_to_4",3,"female","none",3,"real_estate",29,"none","own",1,"skilled",1,"none","yes"],
  ["0_to_200",4,"no_credits","car_new",250,"less_100","unemployed",2,"female","none",3,"real_estate",23,"none","rent",1,"management_self-employed",1,"none","yes"],
  ["no_checking",17,"credits_paid_to_date","car_new",832,"100_to_500","1_to_4",2,"male","none",2,"real_estate",42,"none","own",1,"skilled",1,"none","yes"],
  ["no_checking",33,"outstanding_credit","appliances",5696,"unknown","greater_7",4,"male","co-applicant",4,"unknown",54,"none","free",2,"skilled",1,"yes","yes"],
  ["0_to_200",13,"prior_payments_delayed","retraining",1375,"100_to_500","4_to_7",3,"male","none",3,"real_estate",37,"none","own",2,"management_self-employed",1,"none","yes"]
]

payload_scoring = {"input_data": [{"fields": fields, "values": values}]}
predictions = wml_client.deployments.score(deployment_uid, payload_scoring)

print("Single record scoring result:", "\n fields:", predictions["predictions"][0]["fields"], "\n values: ", predictions["predictions"][0]["values"][0])

Single record scoring result: 
 fields: ['CheckingStatus', 'LoanDuration', 'CreditHistory', 'LoanPurpose', 'LoanAmount', 'ExistingSavings', 'EmploymentDuration', 'InstallmentPercent', 'Sex', 'OthersOnLoan', 'CurrentResidenceDuration', 'OwnsProperty', 'Age', 'InstallmentPlans', 'Housing', 'ExistingCreditsCount', 'Job', 'Dependents', 'Telephone', 'ForeignWorker', 'CheckingStatus_IX', 'CreditHistory_IX', 'LoanPurpose_IX', 'ExistingSavings_IX', 'EmploymentDuration_IX', 'Sex_IX', 'OthersOnLoan_IX', 'OwnsProperty_IX', 'InstallmentPlans_IX', 'Housing_IX', 'Job_IX', 'Telephone_IX', 'ForeignWorker_IX', 'features', 'rawPrediction', 'probability', 'prediction', 'predictedLabel'] 
 values:  ['no_checking', 13, 'credits_paid_to_date', 'car_new', 1343, '100_to_500', '1_to_4', 2, 'female', 'none', 3, 'savings_insurance', 46, 'none', 'own', 2, 'skilled', 1, 'none', 'yes', 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, [20, [1, 3, 5, 13, 14, 15, 16, 17, 18, 19], [1.0, 1.0, 1.0, 13.0, 

## Check if WML payload logging worked else manually store payload records

In [ ]:
import uuid
from ibm_watson_openscale.supporting_classes.payload_record import PayloadRecord
time.sleep(5)
pl_records_count = wos_client.data_sets.get_records_count(payload_data_set_id)
print("Number of records in the payload logging table: {}".format(pl_records_count))
if pl_records_count == 0:
    print("Payload logging did not happen, performing explicit payload logging.")
    wos_client.data_sets.store_records(data_set_id=payload_data_set_id, request_body=[PayloadRecord(
                   scoring_id=str(uuid.uuid4()),
                   request=payload_scoring,
                   response={"fields": predictions['predictions'][0]['fields'], "values":predictions['predictions'][0]['values']},
                   response_time=460
               )],background_mode=False)
    time.sleep(5)
    pl_records_count = wos_client.data_sets.get_records_count(payload_data_set_id)
    print("Number of records in the payload logging table: {}".format(pl_records_count))

Number of records in the payload logging table: 0
Payload logging did not happen, performing explicit payload logging.



 Waiting for end of storing records with request id: 73419321-4efd-ba28-cfc6-f68e85543df9 




active

---------------------------------------
 Successfully finished storing records 
---------------------------------------


Number of records in the payload logging table: 8


# Quality monitoring and feedback logging <a name="quality"></a>

## Enable quality monitoring

The code below waits ten seconds to allow the payload logging table to be set up before it begins enabling monitors. First, it turns on the quality (accuracy) monitor and sets an alert threshold of 70%. OpenScale will show an alert on the dashboard if the model accuracy measurement (area under the curve, in the case of a binary classifier) falls below this threshold.

The second paramater supplied, min_records, specifies the minimum number of feedback records OpenScale needs before it calculates a new measurement. The quality monitor runs hourly, but the accuracy reading in the dashboard will not change until an additional 50 feedback records have been added, via the user interface, the Python client, or the supplied feedback endpoint.

In [ ]:
import time

time.sleep(10)
target = Target(
        target_type=TargetTypes.SUBSCRIPTION,
        target_id=subscription_id
)
parameters = {
    "min_feedback_data_size": 50
}
thresholds = [
                {
                    "metric_id": "area_under_roc",
                    "type": "lower_limit",
                    "value": .80
                }
            ]
quality_monitor_details = wos_client.monitor_instances.create(
    data_mart_id=data_mart_id,
    background_mode=False,
    monitor_definition_id=wos_client.monitor_definitions.MONITORS.QUALITY.ID,
    target=target,
    parameters=parameters,
    thresholds=thresholds
).result




 Waiting for end of monitor instance creation d1b284ee-e9c2-26fc-7074-8752c3b174bf 




preparing
active

---------------------------------------
 Monitor instance successfully created 
---------------------------------------




In [ ]:
quality_monitor_instance_id = quality_monitor_details.metadata.id
quality_monitor_instance_id

'd1b284ee-e9c2-26fc-7074-8752c3b174bf'

## Feedback logging

The code below downloads and stores enough feedback data to meet the minimum threshold so that OpenScale can calculate a new accuracy measurement. It then kicks off the accuracy monitor. The monitors run hourly, or can be initiated via the Python API, the REST API, or the graphical user interface.

In [ ]:
!rm additional_feedback_data_v2.json
!wget https://raw.githubusercontent.com/IBM/watson-openscale-samples/main/IBM%20Cloud/WML/assets/data/credit_risk/additional_feedback_data_v2.json

rm: additional_feedback_data_v2.json: No such file or directory
--2024-07-16 11:42:41--  https://raw.githubusercontent.com/IBM/watson-openscale-samples/main/IBM%20Cloud/WML/assets/data/credit_risk/additional_feedback_data_v2.json
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8003::154, 2606:50c0:8000::154, 2606:50c0:8001::154, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8003::154|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 50890 (50K) [text/plain]
Saving to: ‘additional_feedback_data_v2.json’

additional_feedback 100%[===================>]  49.70K  --.-KB/s    in 0.03s   

2024-07-16 11:42:42 (1.92 MB/s) - ‘additional_feedback_data_v2.json’ saved [50890/50890]



### Get feedback logging dataset ID

In [ ]:
feedback_dataset_id = None
feedback_dataset = wos_client.data_sets.list(type=DataSetTypes.FEEDBACK, 
                                                target_target_id=subscription_id, 
                                                target_target_type=TargetTypes.SUBSCRIPTION).result
print(feedback_dataset)
feedback_dataset_id = feedback_dataset.data_sets[0].metadata.id
if feedback_dataset_id is None:
    print("Feedback data set not found. Please check quality monitor status.")

{
  "data_sets": [
    {
      "metadata": {
        "id": "e07cf0fc-c591-624b-d99f-0b508750451c",
        "crn": "crn:v1:bluemix:public:aiopenscale:us-south:a/181ed6cc388f47bd9d862fe066f9cfce:80e6093f-5acf-4eb7-9da6-7ba9bf56a929:data_set:e07cf0fc-c591-624b-d99f-0b508750451c",
        "url": "/v2/data_sets/e07cf0fc-c591-624b-d99f-0b508750451c",
        "created_at": "2024-07-16T06:12:30.424000Z",
        "created_by": "iam-ServiceId-2e5c9fda-38bf-4279-9712-cdb3b6f3a7ad",
        "modified_at": "2024-07-16T06:12:30.693000Z",
        "modified_by": "iam-ServiceId-2e5c9fda-38bf-4279-9712-cdb3b6f3a7ad"
      },
      "entity": {
        "data_mart_id": "80e6093f-5acf-4eb7-9da6-7ba9bf56a929",
        "name": "3a293f0a-318d-ce62-423a-9376b02b966a_feedback",
        "description": "3a293f0a-318d-ce62-423a-9376b02b966a_feedback",
        "type": "feedback",
        "target": {
          "target_type": "subscription",
          "target_id": "3a293f0a-318d-ce62-423a-9376b02b966a"
        },
    

In [ ]:
with open('additional_feedback_data_v2.json') as feedback_file:
    additional_feedback_data = json.load(feedback_file)

In [ ]:
time.sleep(20) # Sleep to make sure that feedback dataset id is in active state
wos_client.data_sets.store_records(feedback_dataset_id, request_body=additional_feedback_data, background_mode=False)




 Waiting for end of storing records with request id: 7f10511e-fbfa-1629-630e-2b47f04f073e 




active

---------------------------------------
 Successfully finished storing records 
---------------------------------------




In [ ]:
wos_client.data_sets.get_records_count(data_set_id=feedback_dataset_id)

98

In [ ]:
run_details = wos_client.monitor_instances.run(monitor_instance_id=quality_monitor_instance_id, background_mode=False).result




 Waiting for end of monitoring run 6cb1ff2e-df87-5280-29aa-dfd4d147e95f 




finished

---------------------------
 Successfully finished run 
---------------------------




In [57]:
wos_client.monitor_instances.show_metrics(monitor_instance_id=quality_monitor_instance_id)

2024-07-16 06:13:19.969000+00:00,true_positive_rate,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.36363636363636365,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,area_under_roc,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.6587412587412588,0.8,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,precision,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.8,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,matthews_correlation_coefficient,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.41672426371926674,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,f1_measure,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.5000000000000001,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,accuracy,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.7551020408163265,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,label_skew,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.6909336273400493,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,gini_coefficient,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.31748251748251755,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,log_loss,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.4493805793027406,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:13:19.969000+00:00,false_positive_rate,3ff7b4ae-77e5-78e0-0683-c0e13d20f2c4,0.046153846153846156,None,None,['model_type:original'],quality,d1b284ee-e9c2-26fc-7074-8752c3b174bf,b58e9764-1dcf-2628-7687-bb75ca505249,subscription,3a293f0a-318d-ce62-423a-9376b02b966a


Note: First 10 records were displayed.


# Fairness, drift monitoring and explanations <a name="fairness"></a>

### Fairness configuration

The code below configures fairness monitoring for our model. It turns on monitoring for two features, Sex and Age. In each case, we must specify:

  * Which model feature to monitor
  * One or more **majority** groups, which are values of that feature that we expect to receive a higher percentage of favorable outcomes
  * One or more **minority** groups, which are values of that feature that we expect to receive a higher percentage of unfavorable outcomes
  * The threshold at which we would like OpenScale to display an alert if the fairness measurement falls below (in this case, 95%)

Additionally, we must specify which outcomes from the model are favourable outcomes, and which are unfavourable. We must also provide the number of records OpenScale will use to calculate the fairness score. In this case, OpenScale's fairness monitor will run hourly, but will not calculate a new fairness rating until at least 200 records have been added. Finally, to calculate fairness, OpenScale must perform some calculations on the training data, so we provide the dataframe containing the data.

In [58]:
wos_client.monitor_instances.show()

80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,3a293f0a-318d-ce62-423a-9376b02b966a,subscription,model_health,2024-07-16 06:11:34.218000+00:00,777266b9-4ee6-3bca-141c-00230bc48bf8
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,3a293f0a-318d-ce62-423a-9376b02b966a,subscription,quality,2024-07-16 06:12:29.680000+00:00,d1b284ee-e9c2-26fc-7074-8752c3b174bf
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,3a293f0a-318d-ce62-423a-9376b02b966a,subscription,performance,2024-07-16 06:11:36.282000+00:00,6efaf61d-9362-e967-af7a-8328121f3264
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,e223ca0c-e6f4-5074-21a3-1a3b986ff0eb,subscription,mrm,2024-07-05 14:08:34.802000+00:00,e6b08895-0e4b-4d23-c456-80ef9b951d27
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,695a04bf-402a-31c2-6a5d-6a2ff4a350c4,subscription,model_health,2024-06-06 06:04:08.111000+00:00,1550e60f-dc09-3b9e-4009-5d2d0b950761
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,686e4ec1-2b12-cd71-ba08-9e1087124a7b,subscription,mrm,2024-07-02 10:04:06.083000+00:00,9b7a1b18-4b7d-af98-50a8-d099022d53fd
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,e223ca0c-e6f4-5074-21a3-1a3b986ff0eb,subscription,model_health,2024-07-05 13:57:02.336000+00:00,7d36776c-f86d-cb46-3777-ef027bc844ed
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,07d82ce8-8936-4a13-3261-392d9239aa04,subscription,drift,2024-07-15 11:31:15.872000+00:00,e7ce7ee3-e236-a4fc-a7af-5142bb6842df
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,8dd6a05c-5ea3-8e9b-0d0c-08e200fc9dbc,subscription,model_health,2024-07-11 09:30:12.810000+00:00,2382d841-c629-c8b7-b634-c510c66704a5
80e6093f-5acf-4eb7-9da6-7ba9bf56a929,active,686e4ec1-2b12-cd71-ba08-9e1087124a7b,subscription,model_health,2024-06-02 08:26:53.650000+00:00,c0ae0c16-6da5-6d41-e8c1-3ff33ee267ff


Note: First 10 records were displayed.


In [59]:
target = Target(
    target_type=TargetTypes.SUBSCRIPTION,
    target_id=subscription_id

)
parameters = {
    "features": [
        {"feature": "Sex",
         "majority": ['male'],
         "minority": ['female'],
         "threshold": 0.95
         },
        {"feature": "Age",
         "majority": [[26, 75]],
         "minority": [[18, 25]],
         "threshold": 0.95
         }
    ],
    "favourable_class": ["No Risk"],
    "unfavourable_class": ["Risk"],
    "min_records": 100
}

fairness_monitor_details = wos_client.monitor_instances.create(
    data_mart_id=data_mart_id,
    background_mode=False,
    monitor_definition_id=wos_client.monitor_definitions.MONITORS.FAIRNESS.ID,
    target=target,
    parameters=parameters).result
fairness_monitor_instance_id =fairness_monitor_details.metadata.id
fairness_monitor_instance_id




 Waiting for end of monitor instance creation dac44a34-72e2-e747-211b-fd2d9d4f2ec1 




active

---------------------------------------
 Monitor instance successfully created 
---------------------------------------




'dac44a34-72e2-e747-211b-fd2d9d4f2ec1'

### Drift V2 Configuration

In [61]:
monitor_instances = wos_client.monitor_instances.list().result.monitor_instances
for monitor_instance in monitor_instances:
    monitor_def_id=monitor_instance.entity.monitor_definition_id
    if monitor_def_id == "drift_v2" and monitor_instance.entity.target.target_id == subscription_id:
        wos_client.monitor_instances.delete(monitor_instance.metadata.id)
        print('Deleted existing drift v2 monitor instance with id: ', monitor_instance.metadata.id)


#### Finding feature importances

This is done because we will deduce the important and most important features to the model. This will help in narrowing down the analysis of Drift evaluation in the UI

In [ ]:
feature_importance = dict(zip(features, model_rf.stages[-3].featureImportances))

In [62]:
target = Target(
    target_type=TargetTypes.SUBSCRIPTION,
    target_id=subscription_id
)

parameters = {
        "min_samples": 100,
        "max_samples": 1000,
        "train_archive": True,
        "features": {
            "fields": features,
            "importances": feature_importance
        }
    }
drift_v2_monitor_details = wos_client.monitor_instances.create(
    data_mart_id=data_mart_id,
    monitor_definition_id=wos_client.monitor_definitions.MONITORS.DRIFT_V2.ID,
    target=target,
    parameters=parameters
).result

drift_v2_monitor_instance_id = drift_v2_monitor_details.metadata.id
drift_v2_monitor_instance_id

'24bcca7d-5a22-b834-321a-ebee1f4fb11f'

## Score the model again now that monitoring is configured

This next section randomly selects 200 records from the data feed and sends those records to the model for predictions. This is enough to exceed the minimum threshold for records set in the previous section, which allows OpenScale to begin calculating fairness.

In [63]:
with io.capture_output() as captured:
    !wget https://raw.githubusercontent.com/IBM/watson-openscale-samples/main/IBM%20Cloud/WML/assets/data/credit_risk/german_credit_feed.json
!ls -lh german_credit_feed.json

-rw-r--r--  1 nelwin  staff   2.9M Jul 16 11:48 german_credit_feed.json


Score 200 randomly chosen records

In [64]:
import random

with open('german_credit_feed.json', 'r') as scoring_file:
    scoring_data = json.load(scoring_file)

fields = scoring_data['fields']
values = []
for _ in range(200):
    values.append(random.choice(scoring_data['values']))
payload_scoring = {"input_data": [{"fields": fields, "values": values}]}

scoring_response = wml_client.deployments.score(deployment_uid, payload_scoring)

time.sleep(5)
pl_records_count = wos_client.data_sets.get_records_count(payload_data_set_id)

if pl_records_count == 8:
    print("Payload logging did not happen, performing explicit payload logging.")
    wos_client.data_sets.store_records(data_set_id=payload_data_set_id, request_body=[PayloadRecord(
                   scoring_id=str(uuid.uuid4()),
                   request=payload_scoring,
                   response={"fields": scoring_response['predictions'][0]['fields'], "values":scoring_response['predictions'][0]['values']},
                   response_time=460
               )])
    time.sleep(5)
    pl_records_count = wos_client.data_sets.get_records_count(payload_data_set_id)
    print("Number of records in the payload logging table: {}".format(pl_records_count))

Payload logging did not happen, performing explicit payload logging.
Number of records in payload logging table:  208


**Note:** Now in payload table should be total 208 records.

In [65]:
print('Number of records in payload table: ', wos_client.data_sets.get_records_count(data_set_id=payload_data_set_id))

Number of records in payload table:  208


## Run fairness monitor

Kick off a fairness monitor run on current data. The monitor runs hourly, but can be manually initiated using the Python client, the REST API, or the graphical user interface.

In [66]:
run_details = wos_client.monitor_instances.run(monitor_instance_id=fairness_monitor_instance_id, background_mode=False)




 Waiting for end of monitoring run dac44a34-72e2-e747-211b-fd2d9d4f2ec1 




finished

---------------------------
 Successfully finished run 
---------------------------




In [67]:
time.sleep(10)

wos_client.monitor_instances.show_metrics(monitor_instance_id=fairness_monitor_instance_id)

### Run Drift V2 monitor


Kick off a drift v2 monitor run on current data. The monitor runs every day, but can be manually initiated using the Python client, the REST API.

In [70]:
drift_v2_run_details = wos_client.monitor_instances.run(monitor_instance_id=drift_v2_monitor_instance_id, background_mode=False)




 Waiting for end of monitoring run e5a4e4ba-8264-50d6-76b8-e926231249c8 




running................................................
finished

---------------------------
 Successfully finished run 
---------------------------




In [71]:
time.sleep(15)

wos_client.monitor_instances.show_metrics(monitor_instance_id=drift_v2_monitor_instance_id)

2024-07-16 06:19:52.904904+00:00,records_processed,496d1b71-10cb-8540-c90e-1f0033d128ea,208.0,None,None,"['algorithm_used:total_variation', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for No Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:52.904904+00:00,confidence_drift_score,496d1b71-10cb-8540-c90e-1f0033d128ea,0.1862,None,0.05,"['algorithm_used:total_variation', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for No Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:52.904904+00:00,records_processed,496d1b71-10cb-8540-c90e-1f0033d128ea,208.0,None,None,"['algorithm_used:overlap_coefficient', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for No Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:52.904904+00:00,confidence_drift_score,496d1b71-10cb-8540-c90e-1f0033d128ea,0.1868,None,0.05,"['algorithm_used:overlap_coefficient', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for No Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:58.188821+00:00,records_processed,7a493cf6-5c05-fbb5-251d-d87e5c99d5be,208.0,None,None,"['algorithm_used:total_variation', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:58.188821+00:00,confidence_drift_score,7a493cf6-5c05-fbb5-251d-d87e5c99d5be,0.1657,None,0.05,"['algorithm_used:total_variation', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:58.188821+00:00,records_processed,7a493cf6-5c05-fbb5-251d-d87e5c99d5be,208.0,None,None,"['algorithm_used:overlap_coefficient', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:58.188821+00:00,confidence_drift_score,7a493cf6-5c05-fbb5-251d-d87e5c99d5be,0.1667,None,0.05,"['algorithm_used:overlap_coefficient', 'computed_on:payload', 'field_type:class', 'field_name:Class Probability for Risk']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:58.596595+00:00,records_processed,2b002556-cd10-f364-196e-71f239694e57,208.0,None,None,"['algorithm_used:jensen_shannon', 'computed_on:payload', 'field_type:class', 'field_name:predictedLabel']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a
2024-07-16 06:19:58.596595+00:00,prediction_drift_score,2b002556-cd10-f364-196e-71f239694e57,0.0469,None,0.05,"['algorithm_used:jensen_shannon', 'computed_on:payload', 'field_type:class', 'field_name:predictedLabel']",drift_v2,24bcca7d-5a22-b834-321a-ebee1f4fb11f,e5a4e4ba-8264-50d6-76b8-e926231249c8,subscription,3a293f0a-318d-ce62-423a-9376b02b966a


Note: First 10 records were displayed.


## Configure Explainability

Finally, we provide OpenScale with the training data to enable and configure the explainability features.

In [72]:
target = Target(
    target_type=TargetTypes.SUBSCRIPTION,
    target_id=subscription_id
)
parameters = {
    "enabled": True
}
explainability_details = wos_client.monitor_instances.create(
    data_mart_id=data_mart_id,
    background_mode=False,
    monitor_definition_id=wos_client.monitor_definitions.MONITORS.EXPLAINABILITY.ID,
    target=target,
    parameters=parameters
).result

explainability_monitor_id = explainability_details.metadata.id




 Waiting for end of monitor instance creation 379ad386-eb93-39ae-c01c-a71ac6ac14b3 




preparing..
active

---------------------------------------
 Monitor instance successfully created 
---------------------------------------




## Run explanation for sample record

In [73]:
pl_records_resp = wos_client.data_sets.get_list_of_records(data_set_id=payload_data_set_id, limit=1, offset=0).result
scoring_ids = [pl_records_resp["records"][0]["entity"]["values"]["scoring_id"]]
print("Running explanations on scoring IDs: {}".format(scoring_ids))
explanation_types = ["lime", "contrastive"]
result = wos_client.monitor_instances.explanation_tasks(scoring_ids=scoring_ids, explanation_types=explanation_types, subscription_id=subscription_id).result
print(result)

Running explanations on scoring IDs: ['7512f474-6c88-44bd-a591-24350774b1ad-1']
{
  "metadata": {
    "explanation_task_ids": [
      "5de3ab1d-3cb6-0488-cec3-43a2399c72a9"
    ],
    "created_by": "IBMid-662005298W",
    "created_at": "2024-07-16T06:24:59.486239Z"
  }
}


# Custom monitors and metrics <a name="custom"></a>

## Register custom monitor

In [74]:
def get_definition(monitor_name):
    monitor_definitions = wos_client.monitor_definitions.list().result.monitor_definitions
    
    for definition in monitor_definitions:
        if monitor_name == definition.entity.name:
            return definition
    
    return None

In [75]:
monitor_name = 'my model performance'
metrics = [MonitorMetricRequest(name='sensitivity',
                                thresholds=[MetricThreshold(type=MetricThresholdTypes.LOWER_LIMIT, default=0.8)]),
          MonitorMetricRequest(name='specificity',
                                thresholds=[MetricThreshold(type=MetricThresholdTypes.LOWER_LIMIT, default=0.75)])]
tags = [MonitorTagRequest(name='region', description='customer geographical region')]

existing_definition = get_definition(monitor_name)

if existing_definition is None:
    custom_monitor_details = wos_client.monitor_definitions.add(name=monitor_name, metrics=metrics, tags=tags, background_mode=False).result
else:
    custom_monitor_details = existing_definition




 Waiting for end of adding monitor definition my_model_performance 




finished

-------------------------------------------------
 Successfully finished adding monitor definition 
-------------------------------------------------




### Show available monitors types

In [76]:
wos_client.monitor_definitions.show()

my_model_performance,my model performance,"['sensitivity', 'specificity']"
fairness,Fairness,"['Fairness value', 'Average odds difference', 'False discovery rate difference', 'Error rate difference', 'False negative rate difference', 'False omission rate difference', 'False positive rate difference', 'True positive rate difference', 'Average absolute odds difference', 'Statistical parity difference', 'Impact score']"
model_health,Model health,"['Total scoring requests', 'Total records', 'Average records', 'Median records', 'Maximum records', 'Minimum records', 'Total payload size', 'Average payload size', 'Median payload size', 'Minimum payload size', 'Maximum payload size', 'Average API throughput', 'Minimum API throughput', 'Maximum API throughput', 'Median API throughput', 'Average API latency', 'Minimum API latency', 'Maximum API latency', 'Median API latency', 'Average record throughput', 'Minimum record throughput', 'Maximum record throughput', 'Median record throughput', 'Average record latency', 'Minimum record latency', 'Maximum record latency', 'Median record latency', 'Users', 'Errors', 'Data errors', 'System errors', 'Total input token count', 'Average input token count', 'Median input token count', 'Maximum input token count', 'Minimum input token count', 'Total output token count', 'Average output token count', 'Median output token count', 'Maximum output token count', 'Minimum output token count']"
data_health,Data Health,"['Absence Count', 'Empty Strings', 'Data Type Mismatch', 'Class Confusion', 'Duplicate Rows', 'Unique Columns']"
performance,Performance,['Number of records']
explainability,Explainability,['Global explanation stability']
mrm,Model risk management monitoring,"['Tests run', 'Tests passed', 'Tests failed', 'Tests skipped', 'Fairness score', 'Quality score', 'Drift score']"
generative_ai_quality,Generative AI Quality,"['Records processed', 'ROUGE-1', 'ROUGE-2', 'ROUGE-L', 'ROUGE-Lsum', 'SARI', 'METEOR', 'F1 Score', 'Precision', 'Recall', 'BLEU', 'Jaccard similarity', 'Cosine similarity', 'Output data PII', 'Input data PII', 'Output data HAP', 'Input data HAP', 'Readability', 'Micro F1 Score', 'Macro F1 Score', 'Micro precision', 'Macro precision', 'Micro recall', 'Macro recall', 'Exact match', 'Faithfulness', 'Context relevance', 'Answer relevance', 'Unsuccessful requests']"
drift_v2,Drift v2,"['Output drift', 'Prediction drift', 'Model quality drift', 'Feature drift', 'Input metadata drift', 'Output metadata drift', 'Records processed', 'New records processed']"
drift,Drift,"['Drop in accuracy', 'Predicted accuracy', 'Drop in data consistency']"


Note: First 10 records were displayed.


### Get monitors uids and details

In [77]:
custom_monitor_id = custom_monitor_details.metadata.id

print(custom_monitor_id)

my_model_performance


In [78]:
custom_monitor_details = wos_client.monitor_definitions.get(monitor_definition_id=custom_monitor_id).result
print('Monitor definition details:', custom_monitor_details)

Monitor definition details: {
  "metadata": {
    "id": "my_model_performance",
    "crn": "crn:v1:bluemix:public:aiopenscale:us-south:a/0a3c25959fab4ecea2768fa6b8d61595:80e6093f-5acf-4eb7-9da6-7ba9bf56a929:monitor_definition:my_model_performance",
    "url": "/v2/monitor_definitions/my_model_performance",
    "created_at": "2024-07-16T06:25:10.852000Z",
    "created_by": "IBMid-662005298W"
  },
  "entity": {
    "name": "my model performance",
    "metrics": [
      {
        "name": "sensitivity",
        "thresholds": [
          {
            "type": "lower_limit",
            "default": 0.8
          }
        ],
        "expected_direction": "increasing",
        "id": "sensitivity"
      },
      {
        "name": "specificity",
        "thresholds": [
          {
            "type": "lower_limit",
            "default": 0.75
          }
        ],
        "expected_direction": "increasing",
        "id": "specificity"
      }
    ],
    "tags": [
      {
        "name": "region

## Enable custom monitor for subscription

In [79]:
target = Target(
        target_type=TargetTypes.SUBSCRIPTION,
        target_id=subscription_id
    )

thresholds = [MetricThresholdOverride(metric_id='sensitivity', type = MetricThresholdTypes.LOWER_LIMIT, value=0.9)]

custom_monitor_instance_details = wos_client.monitor_instances.create(
            data_mart_id=data_mart_id,
            background_mode=False,
            monitor_definition_id=custom_monitor_id,
            target=target,
            parameters={}
).result




 Waiting for end of monitor instance creation 8b421e8f-3f85-f197-1536-bf14ac8cf013 




active

---------------------------------------
 Monitor instance successfully created 
---------------------------------------




### Get monitor instance id and configuration details

In [80]:
custom_monitor_instance_id = custom_monitor_instance_details.metadata.id

In [81]:
custom_monitor_instance_details = wos_client.monitor_instances.get(custom_monitor_instance_id).result
print(custom_monitor_instance_details)

{
  "metadata": {
    "id": "8b421e8f-3f85-f197-1536-bf14ac8cf013",
    "crn": "crn:v1:bluemix:public:aiopenscale:us-south:a/0a3c25959fab4ecea2768fa6b8d61595:80e6093f-5acf-4eb7-9da6-7ba9bf56a929:monitor_instance:8b421e8f-3f85-f197-1536-bf14ac8cf013",
    "url": "/v2/monitor_instances/8b421e8f-3f85-f197-1536-bf14ac8cf013",
    "created_at": "2024-07-16T06:25:46.655000Z",
    "created_by": "IBMid-662005298W",
    "modified_at": "2024-07-16T06:25:46.786000Z",
    "modified_by": "IBMid-662005298W"
  },
  "entity": {
    "data_mart_id": "80e6093f-5acf-4eb7-9da6-7ba9bf56a929",
    "monitor_definition_id": "my_model_performance",
    "target": {
      "target_type": "subscription",
      "target_id": "3a293f0a-318d-ce62-423a-9376b02b966a"
    },
    "parameters": {},
    "status": {
      "state": "active"
    }
  }
}


## Storing custom metrics

In [82]:
from datetime import datetime, timezone, timedelta
from ibm_watson_openscale.base_classes.watson_open_scale_v2 import MonitorMeasurementRequest
measurement_request = [MonitorMeasurementRequest(timestamp=datetime.now(timezone.utc), 
                                                 metrics=[{"specificity": 0.78, "sensitivity": 0.67, "region": "us-south"}])]
print(measurement_request[0])

{
  "timestamp": "2024-07-16T06:25:53.155818Z",
  "metrics": [
    {
      "specificity": 0.78,
      "sensitivity": 0.67,
      "region": "us-south"
    }
  ]
}


In [83]:
published_measurement_response = wos_client.monitor_instances.measurements.add(
    monitor_instance_id=custom_monitor_instance_id,
    monitor_measurement_request=measurement_request).result
published_measurement_id = published_measurement_response[0]["measurement_id"]
print(published_measurement_response)

[{'measurement_id': '7914116c-fa9f-86e6-2c88-b278150388ce', 'metrics': [{'region': 'us-south', 'sensitivity': 0.67, 'specificity': 0.78}], 'timestamp': '2024-07-16T06:25:53.155818Z'}]


### List and get custom metrics

In [84]:
time.sleep(5)
published_measurement = wos_client.monitor_instances.measurements.get(monitor_instance_id=custom_monitor_instance_id, measurement_id=published_measurement_id).result
print(published_measurement)

{
  "metadata": {
    "id": "7914116c-fa9f-86e6-2c88-b278150388ce",
    "crn": "crn:v1:bluemix:public:aiopenscale:us-south:a/0a3c25959fab4ecea2768fa6b8d61595:80e6093f-5acf-4eb7-9da6-7ba9bf56a929:measurement:7914116c-fa9f-86e6-2c88-b278150388ce",
    "url": "/v2/monitor_instances/8b421e8f-3f85-f197-1536-bf14ac8cf013/measurements/7914116c-fa9f-86e6-2c88-b278150388ce",
    "created_at": "2024-07-16T06:25:56.829050Z",
    "created_by": "IBMid-662005298W"
  },
  "entity": {
    "timestamp": "2024-07-16T06:25:53.155818Z",
    "values": [
      {
        "metrics": [
          {
            "id": "sensitivity",
            "value": 0.67
          },
          {
            "id": "specificity",
            "value": 0.78
          }
        ],
        "tags": [
          {
            "id": "region",
            "value": "us-south"
          }
        ]
      }
    ],
    "issue_count": 0,
    "target": {
      "target_type": "subscription",
      "target_id": "3a293f0a-318d-ce62-423a-9376b02b9

# Historical data <a name="historical"></a>

In [85]:
historyDays = 7

## Insert historical fairness metrics

In [86]:
!rm history_fairness_v2.json
with io.capture_output() as captured:
    !wget https://raw.githubusercontent.com/IBM/watson-openscale-samples/main/IBM%20Cloud/WML/assets/data/historical_data/credit_risk/history_fairness_v2.json
!ls -lh history_fairness_v2.json

rm: history_fairness_v2.json: No such file or directory
-rw-r--r--  1 nelwin  staff   1.1M Jul 16 11:56 history_fairness_v2.json


In [87]:
from datetime import datetime, timedelta, timezone

from ibm_watson_openscale.base_classes.watson_open_scale_v2 import Source
from ibm_watson_openscale.base_classes.watson_open_scale_v2 import Measurements
with open("history_fairness_v2.json") as f:
    fairness_values = json.load(f)
    for day in range(historyDays):
        print('Loading day', day + 1)
        daily_measurement_requests = []
        sources_list = []
        for hour in range(24):
            score_time = (datetime.now(timezone.utc) + timedelta(hours=(-(24*day + hour + 1)))).strftime('%Y-%m-%dT%H:%M:%SZ')
            index = (day * 24 + hour) % len(fairness_values) # wrap around and reuse values if needed
            fairness_values[index]["timestamp"] = score_time
            #print(score_time) 
            fairness_value = fairness_values[index]
            metrics_list = fairness_value["metrics"]
            sources = fairness_value["sources"]
            sources_list = []
            for source in sources:
                source_id = source["id"]
                source_type = source["type"]
                source_data = source["data"]
                if source_id == "bias_detection_summary":
                    source_data["evaluated_at"] = score_time
                    source_data["favourable_class"] = ["No Risk"]
                    source_data["unfavourable_class"] = ["Risk"]
                    source_data["score_type"] = "disparate impact"
                sources_list.append(
                    Source(
                        id=source_id,
                        type=source_type,
                        data=source_data
                    )
                )  
            measurement_request = MonitorMeasurementRequest(metrics=metrics_list, sources=sources_list, timestamp=score_time)
            daily_measurement_requests.append(measurement_request)
        measurements_client = Measurements(wos_client)
        measurements_client.add(monitor_instance_id=fairness_monitor_instance_id, monitor_measurement_request=daily_measurement_requests)     
print('Finished')

Loading day 1
Loading day 2
Loading day 3
Loading day 4
Loading day 5
Loading day 6
Loading day 7
Finished


## Insert historical debias metrics

In [88]:
!rm history_debias_v2.json
with io.capture_output() as captured:
    !wget https://raw.githubusercontent.com/IBM/watson-openscale-samples/main/IBM%20Cloud/WML/assets/data/historical_data/credit_risk/history_debias_v2.json
!ls -lh history_debias_v2.json

rm: history_debias_v2.json: No such file or directory
-rw-r--r--  1 nelwin  staff   725K Jul 16 11:56 history_debias_v2.json


In [89]:
with open("history_debias_v2.json") as f:
    debias_values = json.load(f)
    for day in range(historyDays):
        print('Loading day', day + 1)
        daily_measurement_requests = []
        sources_list = []
        for hour in range(24):
            score_time = (datetime.now(timezone.utc) + timedelta(hours=(-(24*day + hour + 1)))).strftime('%Y-%m-%dT%H:%M:%SZ')
            index = (day * 24 + hour) % len(debias_values) # wrap around and reuse values if needed
            debias_values[index]["timestamp"] = score_time
            debias_value = debias_values[index]
            metrics_list = debias_value["metrics"]
            sources = debias_value["sources"]
            sources_list = []
            for source in sources:
                sources_list.append(
                    Source(
                        id=source["id"],
                        type=source["type"],
                        data=source["data"]
                    )
                )  
            measurement_request = MonitorMeasurementRequest(metrics=metrics_list, sources=sources_list, timestamp=score_time)
            daily_measurement_requests.append(measurement_request)
        measurements_client = Measurements(wos_client)
        measurements_client.add(monitor_instance_id=fairness_monitor_instance_id, monitor_measurement_request=daily_measurement_requests)         

print('Finished')

Loading day 1
Loading day 2
Loading day 3
Loading day 4
Loading day 5
Loading day 6
Loading day 7
Finished


## Insert historical quality metrics

In [90]:
measurements = [0.76, 0.78, 0.68, 0.72, 0.73, 0.77, 0.80]
for day in range(historyDays):
    quality_measurement_requests = []
    print('Loading day', day + 1)
    for hour in range(24):
        score_time = datetime.utcnow() + timedelta(hours=(-(24*day + hour + 1)))
        score_time = score_time.isoformat() + "Z"
        
        metric = {"area_under_roc": measurements[day]}
                
        measurement_request = MonitorMeasurementRequest(timestamp=score_time,metrics = [metric])
        quality_measurement_requests.append(measurement_request)
        
        
    response = wos_client.monitor_instances.measurements.add(
                                            monitor_instance_id=quality_monitor_instance_id,
                                            monitor_measurement_request=quality_measurement_requests).result    
    
print('Finished')

Loading day 1
Loading day 2
Loading day 3
Loading day 4
Loading day 5
Loading day 6
Loading day 7
Finished


## Insert historical confusion matrixes

In [91]:
!rm history_quality_metrics.json
with io.capture_output() as captured:
    !wget https://raw.githubusercontent.com/IBM/watson-openscale-samples/main/IBM%20Cloud/WML/assets/data/historical_data/credit_risk/history_quality_metrics.json
!ls -lh history_quality_metrics.json

rm: history_quality_metrics.json: No such file or directory
-rw-r--r--  1 nelwin  staff    78K Jul 16 11:56 history_quality_metrics.json


In [92]:
from ibm_watson_openscale.base_classes.watson_open_scale_v2 import Source

with open('history_quality_metrics.json') as json_file:
    records = json.load(json_file)
    
for day in range(historyDays):
    index = 0
    cm_measurement_requests = []
    print('Loading day', day + 1)
    
    for hour in range(24):
        score_time = datetime.utcnow() + timedelta(hours=(-(24*day + hour + 1)))
        score_time = score_time.isoformat() + "Z"

        metric = records[index]['metrics']
        source = records[index]['sources']

        
        measurement_request = {"timestamp": score_time, "metrics": [metric], "sources": [source]}
        cm_measurement_requests.append(measurement_request)

        index+=1

    response = wos_client.monitor_instances.measurements.add(monitor_instance_id=quality_monitor_instance_id, monitor_measurement_request=cm_measurement_requests).result    

print('Finished')

Loading day 1
Loading day 2
Loading day 3
Loading day 4
Loading day 5
Loading day 6
Loading day 7
Finished


## Insert historical performance metrics

In [93]:
target = Target(
        target_type=TargetTypes.INSTANCE,
        target_id=payload_data_set_id
    )


performance_monitor_instance_details = wos_client.monitor_instances.create(
            data_mart_id=data_mart_id,
            background_mode=False,
            monitor_definition_id=wos_client.monitor_definitions.MONITORS.PERFORMANCE.ID,
            target=target,
            parameters={}
).result
performance_monitor_instance_id = performance_monitor_instance_details.metadata.id





 Waiting for end of monitor instance creation 9f1be24f-7700-1df3-6f11-a6712dfb68ac 




active

---------------------------------------
 Monitor instance successfully created 
---------------------------------------




In [94]:
for day in range(historyDays):
    performance_measurement_requests = []
    print('Loading day', day + 1)
    for hour in range(24):
        score_time = datetime.utcnow() + timedelta(hours=(-(24*day + hour + 1)))
        score_time = score_time.isoformat() + "Z"
        score_count = random.randint(60, 600)
        
        metric = {"record_count": score_count, "data_set_type": "scoring_payload"}
        
        measurement_request = {"timestamp": score_time, "metrics": [metric]}
        performance_measurement_requests.append(measurement_request)
        
    response = wos_client.monitor_instances.measurements.add(
                                            monitor_instance_id=performance_monitor_instance_id,
                                            monitor_measurement_request=performance_measurement_requests).result    

print('Finished')

Loading day 1
Loading day 2
Loading day 3
Loading day 4
Loading day 5
Loading day 6
Loading day 7
Finished


## Additional data to help debugging

In [97]:
print('Datamart:', data_mart_id)
print('Model:', model_uid)
print('Deployment:', deployment_uid)

Datamart: 80e6093f-5acf-4eb7-9da6-7ba9bf56a929
Model: 857af6d0-2c26-572a-7430-277e506d0372
Deployment: 198d218d-a4d7-88c9-433a-2cbe9b9ef9ef


## Identify transactions for Explainability

Transaction IDs identified by the cells below can be copied and pasted into the Explainability tab of the OpenScale dashboard.

In [98]:
wos_client.data_sets.show_records(payload_data_set_id, limit=5)

None,630,7512f474-6c88-44bd-a591-24350774b1ad-1,1.0,6.0,1.0,2,less_100,2024-07-16T06:18:47.201Z,0.8866628721250828,"[1.0, 3.0, 6.0, 0.0, 3.0, 1.0, 0.0, 0.0, 2.0, 1.0, 1.0, 0.0, 0.0, 10.0, 630.0, 2.0, 2.0, 33.0, 1.0, 1.0]",less_0,vacation,No Risk,0.0,0.0,unskilled,bank,0.0,3.0,none,10,33,all_credits_paid_back,2.0,2,0.0,3.0,1.0,0.0,yes,rent,1.0,1,0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644,"[17.733257442501653, 2.2667425574983446]",none,0.0,female,1,savings_insurance,less_1,"[0.8866628721250828, 0.11333712787491725]"
None,1845,7512f474-6c88-44bd-a591-24350774b1ad-10,0.0,9.0,3.0,2,100_to_500,2024-07-16T06:18:47.201Z,0.8184510088196267,"[3.0, 2.0, 9.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 22.0, 1845.0, 2.0, 2.0, 31.0, 1.0, 1.0]",greater_200,business,No Risk,0.0,0.0,skilled,none,1.0,2.0,none,22,31,outstanding_credit,0.0,2,0.0,1.0,0.0,1.0,no,rent,1.0,1,0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644,"[16.369020176392535, 3.6309798236074644]",none,0.0,male,1,savings_insurance,4_to_7,"[0.8184510088196267, 0.1815489911803732]"
None,3180,7512f474-6c88-44bd-a591-24350774b1ad-100,0.0,0.0,0.0,2,500_to_1000,2024-07-16T06:18:47.201Z,0.707344075368278,"[0.0, 1.0, 0.0, 2.0, 1.0, 0.0, 0.0, 1.0, 1.0, 0.0, 1.0, 0.0, 0.0, 9.0, 3180.0, 2.0, 4.0, 31.0, 1.0, 1.0]",no_checking,car_new,No Risk,1.0,0.0,unskilled,stores,2.0,1.0,none,9,31,credits_paid_to_date,1.0,4,0.0,1.0,1.0,0.0,yes,own,0.0,1,0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644,"[14.146881507365562, 5.853118492634437]",none,0.0,male,1,car_other,4_to_7,"[0.707344075368278, 0.29265592463172185]"
None,311,7512f474-6c88-44bd-a591-24350774b1ad-101,1.0,4.0,3.0,3,500_to_1000,2024-07-16T06:18:47.201Z,0.7874175935565438,"[3.0, 1.0, 4.0, 2.0, 1.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 10.0, 311.0, 3.0, 2.0, 36.0, 1.0, 1.0]",greater_200,appliances,No Risk,1.0,0.0,skilled,none,2.0,1.0,none,10,36,credits_paid_to_date,0.0,2,0.0,1.0,0.0,0.0,yes,rent,1.0,1,0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644,"[15.748351871130874, 4.251648128869124]",none,0.0,female,1,car_other,4_to_7,"[0.7874175935565438, 0.21258240644345622]"
None,9300,7512f474-6c88-44bd-a591-24350774b1ad-102,0.0,9.0,2.0,5,500_to_1000,2024-07-16T06:18:47.201Z,0.578602331772885,"[2.0, 1.0, 9.0, 2.0, 1.0, 0.0, 1.0, 1.0, 0.0, 0.0, 2.0, 1.0, 1.0, 25.0, 9300.0, 5.0, 3.0, 46.0, 3.0, 1.0]",0_to_200,business,Risk,1.0,1.0,management_self-employed,none,2.0,1.0,co-applicant,25,46,credits_paid_to_date,0.0,3,1.0,1.0,2.0,1.0,no,own,0.0,3,0cd1292a-fd8b-fdb9-2ff5-65dd19eb6644,"[8.427953364542299, 11.5720466354577]",yes,1.0,male,1,car_other,4_to_7,"[0.42139766822711494, 0.578602331772885]"


## Congratulations!

You have finished the hands-on lab for IBM Watson OpenScale. You can now view the [OpenScale Dashboard](https://aiopenscale.cloud.ibm.com/). Click on the tile for the German Credit model to see fairness, accuracy, and performance monitors. Click on the timeseries graph to get detailed information on transactions during a specific time window.

## Next steps

OpenScale shows model performance over time. You have two options to keep data flowing to your OpenScale graphs:
  * Download, configure and schedule the [model feed notebook](https://raw.githubusercontent.com/emartensibm/german-credit/master/german_credit_scoring_feed.ipynb). This notebook can be set up with your WML credentials, and scheduled to provide a consistent flow of scoring requests to your model, which will appear in your OpenScale monitors.
  * Re-run this notebook. Running this notebook from the beginning will delete and re-create the model and deployment, and re-create the historical data. Please note that the payload and measurement logs for the previous deployment will continue to be stored in your datamart, and can be deleted if necessary.